### DISCUSSION TITLE
Is there a way to trigger formatter after insert new code by code action?

### DISCUSSION THREAD

**Discussion by jiashengguo** [eclipse-langium/langium#892] [Q&A]
It would be nice if the inserted code could be formatted instead of manually inserting the indent in the code action.

**Comment by msujew** [comment]
Hey @jiashengguo,

I would generally recommend to have some sort of infrastructure in place to generate correctly indented code in the first place. There are a few methods in the [`Formatter`](https://github.com/langium/langium/blob/main/packages/langium/src/lsp/formatter.ts) such as [`getExistingIndentationCharacterCount`](https://github.com/langium/langium/blob/c7e15ee9663d37dab676dea4ffce479e3af0fea3/packages/langium/src/lsp/formatter.ts#L300-L304) which can be used as inspiration to figure out the required indentation at a specific line. In theory you can also utilize the [`TextDocument.applyEdits`](https://github.com/langium/langium/blob/c7e15ee9663d37dab676dea4ffce479e3af0fea3/packages/langium/src/test/langium-test.ts#L249) function and then trigger the formatter, but that leads down a rabbit hole, since you will need to create copies of a document, reparse it, apply the formatting, etc., which can quickly become very complicated.

**Comment by jiashengguo** [comment]
Hi @msujew 
Thanks for your suggestion.  If I want to call the methods in the Formatter such as getExistingIndentationCharacterCount in CodeActionProvider,  how can I get the `context: FormattingContext`?

**Reply by msujew** [reply]
You will need to reimplement it, especially since the `FormattingContext` isn't available, given that this is not a real formatting operation. You'll need to infer whether the user needs tab or whitespace formatting depending on the existing text. I haven't given much thought to this requirement (and the team at Microsoft behind the LSP as well), since I've never encountered code actions which require to add/replace more than a single line or which are depended on indentation.

**Reply by jiashengguo** [reply]
My case is also inserting a single line, but I want to know whether the current editor uses Tab or Space as an indent. As what  `FormattingOptions.insertSpaces `indicates.

**Reply by msujew** [reply]
Yeah, you'll need to look at the existing user code. The language client only sends this information to the server when actually performing a formatting operation (one that's been started by the user).
